elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,743 @@
+      this.strategy = new TreeStrategy(this.minSplit);
+     else 
+      this.strategy = new NaiveStrategy();
+  }
+
+  public OutlierResult run(Database db, Relation<V> relation) {
+
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("HySortOD scores", relation.size(), LOG) : null;
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+
+    int[] W;
+    List<Hypercube> H;
+
+    {
+      H = getSortedHypercubes(relation);
+      W = this.strategy.buildIndex(H).getDensities();
+
+      // Iterate over all objects
+      int objectPosition = 0;
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance())
+        scores.putDouble(iter, score(W[this.invertedIndex[objectPosition++]]));
+
+      LOG.incrementProcessed(prog);
+    }
+
+    LOG.ensureCompleted(prog);
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("HySortOD", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  private List<Hypercube> getSortedHypercubes(Relation<V> relation) {
+    BinaryTree<Hypercube> sorted = new BinaryTree<>();
+
+    int objectPosition = 0;
+    // Iterate over all objects
+    for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance()) {
+      Hypercube h = new Hypercube(relation.get(iter), this.l);
+      Hypercube obj = sorted.get(h);
+
+      if(Objects.isNull(obj)) {
+        h.add(objectPosition);
+        sorted.add(h);
+      }
+      else {
+        obj.add(objectPosition);
+      }
+
+      objectPosition++;
+    }
+
+    int n = sorted.size();
+    List<Hypercube> H = new ArrayList<Hypercube>(n);
+
+    this.invertedIndex = new int[relation.size()];
+    int hypercybePosition = 0;
+    for(Hypercube h : sorted) {
+      H.add(h);
+
+      for(Integer _objectPosition : h.getInstances())
+        this.invertedIndex[_objectPosition] = hypercybePosition;
+
+      hypercybePosition++;
+    }
+
+    return H;
+  }
+
+  private double score(int density) {
+    return 1 - (density / this.strategy.getMaxDensity());
+  }
+
+  @Override
+  public TypeInformation[] getInputTypeRestriction() {
+    return TypeUtil.array(TypeUtil.NUMBER_VECTOR_FIELD);
+  }
+
+  class Hypercube implements Comparable<Hypercube>, Comparator<Hypercube> {
+
+    int[] coords;
+
+    ArrayList<Integer> instances;
+
+    public Hypercube(V instance, double lenght) {
+      super();
+      this.coords = new int[instance.getDimensionality()];
+      this.instances = new ArrayList<Integer>();
```

Comment:
Please use `IntegerArray` with native integers to avoid the boxing overhead of `Integer.